src/librustc_middle/ty/sty.rs

```diff
@@ -2339,7 +2339,21 @@ impl<'tcx> Const<'tcx> {
     /// Tries to evaluate the constant if it is `Unevaluated`. If that doesn't succeed, return the
     /// unevaluated constant.
     pub fn eval(&self, tcx: TyCtxt<'tcx>, param_env: ParamEnv<'tcx>) -> &Const<'tcx> {
-        let try_const_eval = |did, param_env: ParamEnv<'tcx>, substs, promoted| {
+        let try_const_eval = |did, param_env: ParamEnv<'tcx>, substs: SubstsRef<'tcx>, promoted| {
+            // See if we can bail out early without calling 'with_reveal_all_normalized'
+            // The behavior of `ParamEnv.and` is affected by the reveal mode
+            // (which is set by `with_reveal_all_normalized`).
+            // This means that we can't simply call 'param_env.and(substs).has_local_value()',
+            // since `ParamEnv.and` might discard the caller bounds (which might have inference
+            // vars) after we call `with_reveal_all_normalized`.
+            //
+            // To avoid depending on the precise behavior of `ParamEnv.and`, or needing
+            // to manually set the `ParamEnv` reveal mode, we only check the substs.
+            // If the substs have any inference variables, then the `ParamEnvAnd` is guaranteed
+            // to as well, so there's no risk of bailing out incorrectly.
+            if substs.needs_infer() {
+                return None;
+            }
```

Comment:
Might be worth checking out one of my "ty: erase lifetimes early in `ty::Const::eval`." commit from #70452.

IMO it's easier to understand the code after that commit than before, and I think it fixes a latent bug too.

Reply:
But I guess your early check is correct, `param_env.and(substs)` wouldn't remove anything from `substs`.